components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java

```diff
@@ -0,0 +1,203 @@
+        this.onErrorSeq = onErrorSeq;
+        this.sequential = sequential;
+        this.synapseEnvironment = synapseEnvironment;
+    }
+
+    /**
+     * This method will be called when the initiated gRPC call is expecting an response.
+     * This will inject the gRPC message to the to the message sequence as well as the responseObserver .
+     * inject the message.
+     * @param receivedEvent received event from gRPC
+     * @param responseObserver object to be used when sending the response back to the gRPC client
+     */
+    public void invokeProcess(Event receivedEvent, StreamObserver<Event> responseObserver) {
+        try {
+            org.apache.synapse.MessageContext msgCtx = createMessageContext();
+            if (msgCtx != null) {
+                msgCtx.setProperty(InboundEndpointConstants.INBOUND_ENDPOINT_RESPONSE_WORKER,
+                        new GRPCResponseSender(responseObserver));
+            }
+            initiateSequenceAndInjectPayload(responseObserver, receivedEvent, msgCtx);
+        } catch (AxisFault e) {
+            log.error("Error while processing the gRPC Message", e);
+            throw new SynapseException("Error while processing the gRPC Message", e);
+        }
+    }
+
+    /**
+     * This method will be called when the initiated gRPC call is not expecting an response.
+     * This will inject the gRPC message to the to the message sequence as well as the responseObserver .
+     * inject the message.
+     * @param receivedEvent received event from gRPC
+     * @param responseObserver object to be used to send message processing error back to gRPC client
+     */
+    public void invokeConsume(Event receivedEvent, StreamObserver<Empty> responseObserver) {
+        try {
+            initiateSequenceAndInjectPayload(responseObserver, receivedEvent, createMessageContext());
+        } catch (AxisFault e) {
+            log.error("Error while consuming the Grpc Message", e);
+            throw new SynapseException("Error while consuming the JMS Message", e);
+        }
+    }
+
+    private void initiateSequenceAndInjectPayload(StreamObserver responseObserver,
+                                                  Event receivedEvent,
+                                                  org.apache.synapse.MessageContext msgCtx) throws AxisFault {
+        String msgPayload = receivedEvent.getPayload();
+        String sequenceName = receivedEvent.getHeadersMap().get(HEADER_MAP_SEQUENCE_PARAMETER_NAME);
+        SequenceMediator seq;
+        if (sequenceName != null) {
+            if (log.isDebugEnabled()) {
+                log.debug(sequenceName + " sequence, received via gRPC headers.");
+            }
+            seq = (SequenceMediator) synapseEnvironment.getSynapseConfiguration().getSequence(sequenceName);
+        } else {
+            if (injectingSeq == null || injectingSeq.isEmpty()) {
+                log.error("Sequence name is not specified in inbound endpoint or empty.");
```

Comment:
Rather, The sequence + "sequenceName" specified in the inbound endpoint is incorrect?